pkg/testing/ess/provisioner.go

```diff
@@ -62,89 +60,77 @@ func (p *provisioner) SetLogger(l runner.Logger) {
 	p.logger = l
 }
 
-func (p *provisioner) Provision(ctx context.Context, requests []runner.StackRequest) ([]runner.Stack, error) {
-	results := make(map[runner.StackRequest]*CreateDeploymentResponse)
-	for _, r := range requests {
-		// allow up to 2 minutes for each create request
-		createCtx, createCancel := context.WithTimeout(ctx, 2*time.Minute)
-		resp, err := p.createDeployment(createCtx, r,
-			map[string]string{
-				"division":          "engineering",
-				"org":               "ingest",
-				"team":              "elastic-agent",
-				"project":           "elastic-agent",
-				"integration-tests": "true",
-			})
-		createCancel()
-		if err != nil {
-			return nil, err
-		}
-		results[r] = resp
-	}
+// Create creates a stack.
+func (p *provisioner) Create(ctx context.Context, request runner.StackRequest) (runner.Stack, error) {
+	// allow up to 2 minutes for request
+	createCtx, createCancel := context.WithTimeout(ctx, 2*time.Minute)
+	defer createCancel()
+	resp, err := p.createDeployment(createCtx, request,
+		map[string]string{
+			"division":          "engineering",
+			"org":               "ingest",
+			"team":              "elastic-agent",
+			"project":           "elastic-agent",
+			"integration-tests": "true",
+		})
+	if err != nil {
+		return runner.Stack{}, err
+	}
+	return runner.Stack{
+		ID:            request.ID,
+		Version:       request.Version,
+		Elasticsearch: resp.ElasticsearchEndpoint,
+		Kibana:        resp.KibanaEndpoint,
+		Username:      resp.Username,
+		Password:      resp.Password,
+		Internal: map[string]interface{}{
+			"deployment_id": resp.ID,
+		},
+		Ready: false,
+	}, nil
+}
 
-	// set a long timeout
-	// this context travels up to the magefile, clients that want a shorter timeout can set
-	// it via mage's -t flag
-	readyCtx, readyCancel := context.WithTimeout(ctx, 25*time.Minute)
-	defer readyCancel()
-
-	g, gCtx := errgroup.WithContext(readyCtx)
-	for req, resp := range results {
-		g.Go(func(req runner.StackRequest, resp *CreateDeploymentResponse) func() error {
-			return func() error {
-				ready, err := p.client.DeploymentIsReady(gCtx, resp.ID, 30*time.Second)
-				if err != nil {
-					return fmt.Errorf("failed to check for cloud %s to be ready: %w", req.Version, err)
-				}
-				if !ready {
-					return fmt.Errorf("cloud %s never became ready: %w", req.Version, err)
-				}
-				return nil
-			}
-		}(req, resp))
+// WaitForReady should block until the stack is ready or the context is cancelled.
+func (p *provisioner) WaitForReady(ctx context.Context, stack runner.Stack) (runner.Stack, error) {
+	deploymentID, err := p.getDeploymentID(stack)
+	if err != nil {
+		return stack, fmt.Errorf("failed to get deployment ID from the stack: %w", err)
 	}
-	err := g.Wait()
+	// allow up to 10 minutes for it to become ready
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Minute)
```

Comment:
I made it this way allowing the provisioner code to make the decision on what is the best timeout to be used based on what is being used. That way a timeout from a setting or default value doesn't effect the needed time. I believe its the provisioners that should be determine the appropriate amount of time.

Reply:
Well it seems to me that we have at least one occurrence where these timeouts are related 
https://github.com/elastic/elastic-agent/pull/3701/files#diff-488eb64eb90cc8cd9d898511bc73a09b146c6704f3ce57d9e88d0189cd007498R693-R695

or is it just coincidence?

The timeout should be how long the caller is willing to wait, so why can't we pass that in ?

In general I don't see the downside of putting the timeouts as parameters (even if at first we just pass constants in it) rather than hardcoding different timeouts in different functions

Reply:
That timeout places a hard limit just for caution. I could remove it, but if the provisioner doesn't place a timeout on its own execution then it would result in it blocking forever. If you prefer I can remove it here.

Reply:
I like how it is.

Reply:
The caller has the flexibility to set a shorter timeout through the context. If the parent context's timeout is shorter than the child context's the child context expires when the parents does.

So as is you can change the timeout with the context, and both the maximum and default timeouts are 10 minutes. 

https://go.dev/play/p/sQQpd4QWt42

```go
package main

import (
	"context"
	"fmt"
	"time"
)

func main() {
	ctx1, cancel1 := context.WithTimeout(context.Background(), 1*time.Millisecond)
	defer cancel1()
	ctx5, cancel5 := context.WithTimeout(ctx1, 5*time.Second)
	defer cancel5()

	select {
	case <-ctx1.Done():
		fmt.Println("Waited 1 millisecond")
	case <-ctx5.Done():
		fmt.Println("Waited 5 seconds")
	}
}
```

This prints `Waited 1 millisecond`.

Reply:
> So as is you can change the timeout with the context, and both the maximum and default timeouts are 10 minutes.

To be clear I think this is a totally reasonable way for this to work, unless we think the timeout should be able to be longer than 10 minutes.

Passing in a timeout or deadline via the context is a fine way to do this.

Reply:
Updated comment with a correct example:

https://go.dev/play/p/KBEtIjKFho_b

```go
package main

import (
	"context"
	"fmt"
	"time"
)

func main() {
	ctx1, cancel1 := context.WithTimeout(context.Background(), 1*time.Millisecond)
	defer cancel1()
	ctx5, cancel5 := context.WithTimeout(ctx1, 5*time.Second)
	defer cancel5()

	start := time.Now()
	select {
	case <-ctx5.Done():
		fmt.Printf("Child context is expired after %s\n", time.Since(start))
	}
}
```

This prints `Child context is expired after 1ms` since the child context respects the parent context's deadline.

So with the code today the maximum and default timeout is 10 minutes, and callers can set a shorter timeout via the context. I think this is a perfectly reasonable way for this to work.

The only debate should be if 10 minutes is a long enough maximum timeout.